core/toolbox/category.js

```diff
@@ -0,0 +1,826 @@
+  toolboxLabel.textContent = name;
+  Blockly.utils.dom.addClass(toolboxLabel, this.cssConfig_['label']);
+  return toolboxLabel;
+};
+
+/**
+ * Create the dom for all subcategories.
+ * @param {!Array<!Blockly.ToolboxItem>} contents The category contents.
+ * @return {!Element} The div holding all the subcategories.
+ * @protected
+ */
+Blockly.ToolboxCategory.prototype.createSubCategoriesDom_ = function(contents) {
+  var contentsContainer = document.createElement('div');
+  Blockly.utils.dom.addClass(contentsContainer, this.cssConfig_['contents']);
+
+  for (var i = 0; i < contents.length; i++) {
+    var newCategory = contents[i];
+    var dom = newCategory.createDom();
+    contentsContainer.appendChild(dom);
+  }
+  return contentsContainer;
+};
+
+/**
+ * Updates the colour for this category.
+ * @public
+ */
+Blockly.ToolboxCategory.prototype.refreshTheme = function() {
+  this.colour_ = this.getColour_(/** @type {Blockly.utils.toolbox.CategoryJson} **/
+      (this.toolboxItemDef_));
+  this.addColourBorder_(this.colour_);
+};
+
+/**
+ * Add the strip of colour to the toolbox category.
+ * @param {string} colour The category colour.
+ * @protected
+ */
+Blockly.ToolboxCategory.prototype.addColourBorder_ = function(colour) {
+  if (colour) {
+    var border = Blockly.ToolboxCategory.borderWidth + 'px solid ' +
+        (colour || '#ddd');
+    if (this.workspace_.RTL) {
+      this.rowDiv_.style.borderRight = border;
+    } else {
+      this.rowDiv_.style.borderLeft = border;
+    }
+  }
+};
+
+/**
+ * Adds either the colour or the style for a category.
+ * @param {!Blockly.utils.toolbox.CategoryJson} categoryDef The object holding
+ *    information on the category.
+ * @return {string} The hex colour for the category.
+ * @protected
+ */
+Blockly.ToolboxCategory.prototype.getColour_ = function(categoryDef) {
+  var styleName = categoryDef['categorystyle'] || categoryDef['categoryStyle'];
+  var colour = categoryDef['colour'];
+
+  if (colour && styleName) {
+    console.warn('Toolbox category "' + this.name_ +
+        '" must not have both a style and a colour');
+  } else if (styleName) {
+    return this.getColourfromStyle_(styleName);
+  } else {
+    return this.parseColour_(colour);
+  }
+  return '';
+};
+
+/**
+ * Retrieves and sets the colour for the category using the style name.
+ * The category colour is set from the colour style attribute.
+ * @param {string} styleName Name of the style.
+ * @return {string} The hex colour for the category.
+ * @private
+ */
+Blockly.ToolboxCategory.prototype.getColourfromStyle_ = function(styleName) {
+  var theme = this.workspace_.getTheme();
+  if (styleName && theme) {
+    var style = theme.categoryStyles[styleName];
+    if (style && style.colour) {
+      return this.parseColour_(style.colour);
+    } else {
+      console.warn('Style "' + styleName +
+          '" must exist and contain a colour value');
+    }
+  }
+  return '';
+};
+
+/**
+ * Sets the colour on the category.
```

Comment:
This looks like it parses, not sets.

This also doesn't look toolbox-specific. How it it different from other places where we parse colours? I think there's a util function that does all of this except for the wording of the warning.